File: problems/hackerrank/arrays/strangeCode/strangeCode.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

// strangeCounter finds the value at time t for a strange counter.
func strangeCounter(t int64) int64 {
	cycle := int64(3)
	for t > cycle {
		t -= cycle
		cycle <<= 1
	}
	return cycle - t + 1
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16 * 1024 * 1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16 * 1024 * 1024)

	t, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
	checkError(err)

	result := strangeCounter(t)

	fmt.Fprintf(writer, "%d\n", result)

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}

File: problems/hackerrank/arrays/strangeCode/strangeCode.py
#!/bin/python3
# Without Library

import os

def strangeCounter(t):
    """
    Finds the value at time t for the strange counter.
    
    Args:
        t (int): The time at which to find the counter value.
        
    Returns:
        int: The value of the counter at time t.
    """
    cycle = 3
    while t > cycle:
        t -= cycle
        cycle <<= 1
    return cycle - t + 1

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    t = int(input().strip())

    result = strangeCounter(t)

    fptr.write(str(result) + '\n')

    fptr.close()


#!/bin/python3
# With Library

import os
from math import log2

def strangeCounter(t):
    """
    Finds the value at time t for the strange counter using log2.
    
    Args:
        t (int): The time at which to find the counter value.
        
    Returns:
        int: The value of the counter at time t.
    """
    k = int(log2((t + 2) / 3))
    return (3 << k) - (t - ((3 << k) - 2))

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    t = int(input().strip())

    result = strangeCounter(t)

    fptr.write(str(result) + '\n')

    fptr.close()